File: src/main/java/com/pocketempire/units/UnitConfigLoader.java
package com.pocketempire.units;

import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.InputStream;
import java.util.HashMap;
import java.util.Map;

public class UnitConfigLoader {
    private static Map<String, UnitStats> configs = new HashMap<>();

    static {
        loadConfigs();
    }

    private static void loadConfigs() {
        try {
            ObjectMapper mapper = new ObjectMapper();
            InputStream is = UnitConfigLoader.class.getResourceAsStream("/units.json");
            UnitStats[] units = mapper.readValue(is, UnitStats[].class);

            for (UnitStats unit : units) {
                configs.put(unit.getType(), unit);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public static UnitStats getConfig(String type) {
        return configs.get(type);
    }
}

File: src/main/kotlin/com/pocketempire/config/UnitConfigLoader.kt
package com.pocketempire.config

import com.fasterxml.jackson.databind.ObjectMapper
import com.pocketempire.units.UnitStats

object UnitConfigLoader {

    private val configs = HashMap<String, UnitStats>()

    init {
        loadConfigs()
    }

    private fun loadConfigs() {
        val mapper = ObjectMapper()

        val inputStream = requireNotNull(
            UnitConfigLoader::class.java.getResourceAsStream("/units.json")
        ) {
            "Could not find /units.json"
        }

        val units = mapper.readValue(
            inputStream,
            Array<UnitStats>::class.java
        )

        for (unit in units) {
            configs[unit.type] = unit
        }
    }

    @JvmStatic
    fun getConfig(type: String): UnitStats {
        return configs[type]
            ?: throw IllegalArgumentException(
                "Unexpected config type: $type"
            )
    }
}